kernel/posix/pthread_semaphore.c

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <pthread.h>
+
+int sem_destroy(sem_t *semaphore)
+{
+	k_sem_reset(semaphore->sem);
+	return 0;
+}
+
+int sem_getvalue(sem_t *semaphore, int *value)
+{
+	int count;
+
+	count = k_sem_count_get(semaphore->sem);
+	*value = count;
+
+	return 0;
+}
+
+int sem_init(sem_t *semaphore, int pshared, unsigned int value)
+{
+	if ((semaphore->sem->limit) == 0) {
+		return -1;
+	}
+	ARG_UNUSED(pshared);
+	__ASSERT((pshared == 1) || (pshared == 0), "not correct pshared");
+	k_sem_init(semaphore->sem, value, semaphore->sem->limit);
+
+	return 0;
+}
+
+int sem_post(sem_t *semaphore)
+{
+	k_sem_give(semaphore->sem);
+
+	if (semaphore == NULL) {
+		return -1;
+	} else {
+		return 0;
+	}
+}
+
+int sem_timedwait(sem_t *semaphore, struct timespec *abstime)
+{
+	u32_t		timeout;
+	long		msecs = 0;
+	s64_t		current_ms;
+	int		abstime_ms;
+
+	__ASSERT(abstime, "abstime pointer NULL");
+	if ((abstime->tv_sec < 0) || (abstime->tv_nsec >= 1000000000)) {
+		return -1;
+	}
+
+	current_ms = k_uptime_get();
+	abstime_ms = _ts_to_ms(abstime);
+	msecs = abstime_ms - current_ms;
+
+	if (abstime_ms <= current_ms) {
+		timeout = 0;
+	} else {
+		timeout = _ms_to_ticks(msecs);
+	}
+
+	if (!(k_sem_take(semaphore->sem, timeout))) {
+		return 0;
+	} else {
+		return -1;
+	}
+}
+
+
+int sem_trywait(sem_t *semaphore)
+{
+	if (!(k_sem_take(semaphore->sem, K_NO_WAIT))) {
+		return 0;
+	} else {
+		return -1;
```

Comment:
Please see comments about returning errors on errno; read the man pages for the functions you're implementing and set errno appropriately.  If the errno from POSIX does not apply to Zephyr, please document the differences from the standard in the header file.

The reasoning for this is that people will be porting code that might rely on errno to handle the error appropriately. Just returning errno here will leave errno with the last value, that won't make any sense to the caller of these shim functions.

Reply:
Hey @lpereira Thanks a lot. This is really useful. I will push fix for all the remaining comments today.